File: real_benchmarks/data/real_c_rust_pairs/varargs_handling.c
/*
 * Variadic function divergence — inspired by CVE-2013-2028 (nginx format string)
 * and countless format string vulnerabilities.
 *
 * C variadic functions (stdarg.h) have no type safety at the call site.
 * Wrong format specifiers → UB (reads wrong type from va_list).
 * Rust has no variadic functions in safe code; macros handle formatting.
 */
#include <stdarg.h>
#include <stdint.h>

/* Sum n integers via va_list. */
int sum_variadic(int count, ...) {
    va_list args;
    va_start(args, count);
    int total = 0;
    for (int i = 0; i < count; i++) {
        total += va_arg(args, int);  /* UB if caller passes fewer than count args */
    }
    va_end(args);
    return total;
}

/* Type mismatch in va_arg: caller passes float, callee reads int.
   This is a real class of printf format string bugs. */
int read_as_int_from_varargs(int dummy, ...) {
    va_list args;
    va_start(args, dummy);
    int val = va_arg(args, int);  /* UB if caller passed a double/float */
    va_end(args);
    return val;
}

/* Max of n int32_t values. */
int32_t max_variadic(int count, ...) {
    va_list args;
    va_start(args, count);
    int32_t max_val = INT32_MIN;
    for (int i = 0; i < count; i++) {
        int32_t v = va_arg(args, int32_t);
        if (v > max_val) max_val = v;
    }
    va_end(args);
    return max_val;
}

File: real_benchmarks/data/real_c_rust_pairs/varargs_handling.rs
/*
 * Rust translation of variadic function patterns.
 * Rust has no variadic functions in safe code.
 * Slices replace va_list — fully type-safe.
 */

/// Sum: Rust uses a slice — type-safe, bounds-checked.
pub fn sum_variadic(values: &[i32]) -> i32 {
    let mut total: i32 = 0;
    for &v in values {
        total = total.wrapping_add(v);
    }
    total
}

/// No type mismatch possible: Rust enforces types at compile time.
/// There is no equivalent of "read int from float vararg" in safe Rust.
pub fn read_as_int_from_varargs(val: i32) -> i32 {
    val  // type is known at compile time, no UB possible
}

/// Max of slice: type-safe, no varargs.
pub fn max_variadic(values: &[i32]) -> i32 {
    let mut max_val = i32::MIN;
    for &v in values {
        if v > max_val {
            max_val = v;
        }
    }
    max_val
}
